File: src/MaximumProductOfThreeNumbers.java
public class MaximumProductOfThreeNumbers {
    public int maximumProduct(int[] numbers) {
        int positive1 = Integer.MIN_VALUE, positive2 = Integer.MIN_VALUE, positive3 = Integer.MIN_VALUE,
                negative1 = Integer.MAX_VALUE,
                negative2 = Integer.MAX_VALUE;

        for (int number : numbers) {
            if (number > positive1) {
                positive3 = positive2;
                positive2 = positive1;
                positive1= number;
            } else if (number > positive2) {
                positive3 = positive2;
                positive2 = number;
            } else if (number> positive3) {
                positive3 = number;
            }
            if (number < negative2) {
                negative1 = negative2;
                negative2 = number;
            } else if (number < negative1) {
                negative1 = number;
            }
        }

        return Math.max(positive1 * positive2 * positive3, negative1 * negative2 * positive1);
    }
}

File: python/maximum_product_of_three_numbers.py
from typing import List


class Solution:
    def maximumProduct(self, array: List[int]) -> int:
        positive_1, positive_2, positive_3 = (-float('inf'), ) * 3
        negative_1, negative_2 = (float('inf'), ) * 2
        for number in array:
            if number > positive_1:
                positive_3 = positive_2
                positive_2 = positive_1
                positive_1 = number
            elif number > positive_2:
                positive_3 = positive_2
                positive_2 = number
            elif number > positive_3:
                positive_3 = number
            if number < negative_1:
                negative_2 = negative_1
                negative_1 = number
            elif number < negative_2:
                negative_2 = number

        return max(positive_1 * positive_2 * positive_3, negative_1 * negative_2 * positive_1)
